### PR TITLE
Use test tree consistently with dist-git fmf

### PR DESCRIPTION
**Pull request by LecrisUT**
Fixes a regression in #4915 where it would not get the appropriate `ref` when requested using the local git repository.

Incidentally, the non-dist-git approach already behaves as we want it with and without `url`, so we can (should be) using the `test_dir` directly. We can thus just drop the special handling of `dist-git` in the `_fetch_local_repository` also.

### CODE REVIEW

**Comment by vaibhavdaren** [comment]
One [Test Case](https://artifacts.dev.testing-farm.io/9bad69ee-7931-41fa-900f-09c898f993c8/work-discover50hyb10c/plans/features/steps/discover/execute/data/guest/default-0/tests/discover/distgit-3/data/Detect-within-extracted-sources-inner-fmf-root-is-used/output.txt) is still failing.

**Comment by LecrisUT** [comment]
> One [Test Case](https://artifacts.dev.testing-farm.io/9bad69ee-7931-41fa-900f-09c898f993c8/work-discover50hyb10c/plans/features/steps/discover/execute/data/guest/default-0/tests/discover/distgit-3/data/Detect-within-extracted-sources-inner-fmf-root-is-used/output.txt) is still failing.

Welp, this is related to one old nasty bug that I wanted to resolve https://github.com/teemtee/tmt/issues/2097

**Comment by LecrisUT** [comment]
Here is the run tree before

<details>

```console
$ tree -a -I .git
.
├── cleanup
│   └── step.yaml
├── data
│   ├── plan-source-script.sh
│   └── variables.env-default-0
├── discover
│   ├── default-0
│   │   ├── source
│   │   │   ├── demo-1.0-build
│   │   │   ├── demo.spec
│   │   │   ├── .fmf
│   │   │   │   └── version
│   │   │   ├── mock_sources
│   │   │   ├── outsider
│   │   │   ├── sources
│   │   │   ├── SPECPARTS
│   │   │   ├── SRPMS
│   │   │   ├── top_test.fmf
│   │   │   ├── with-tmt-1
│   │   │   │   ├── .fmf
│   │   │   │   │   └── version
│   │   │   │   ├── from-source.txt
│   │   │   │   └── tests
│   │   │   │       └── from-source.fmf
│   │   │   └── with-tmt-1.tgz
│   │   └── tests
│   │       ├── .fmf
│   │       │   └── version
│   │       ├── from-source.txt
│   │       └── tests
│   │           └── from-source.fmf
│   ├── step.yaml
│   └── tests.yaml
├── execute
│   ├── results.yaml
│   └── step.yaml
├── finish
│   └── step.yaml
├── prepare
│   ├── results.yaml
│   └── step.yaml
├── provision
│   ├── default-0
│   │   └── podman-run-environment
│   ├── guests.yaml
│   └── step.yaml
├── report
│   └── step.yaml
└── tree
    ├── demo.spec
    ├── .fmf
    │   └── version
    ├── mock_sources
    ├── sources
    ├── top_test.fmf
    └── with-tmt-1.tgz

24 directories, 33 files
```

</details>

after

<details>

```console
$ tree -a -I .git
.
├── cleanup
│   └── step.yaml
├── data
│   ├── plan-source-script.sh
│   └── variables.env-default-0
├── discover
│   ├── default-0
│   │   ├── source
│   │   │   ├── demo-1.0-build
│   │   │   ├── demo.spec
│   │   │   ├── .fmf
│   │   │   │   └── version
│   │   │   ├── mock_sources
│   │   │   ├── outsider
│   │   │   ├── sources
│   │   │   ├── SPECPARTS
│   │   │   ├── SRPMS
│   │   │   ├── top_test.fmf
│   │   │   ├── with-tmt-1
│   │   │   │   ├── .fmf
│   │   │   │   │   └── version
│   │   │   │   ├── from-source.txt
│   │   │   │   └── tests
│   │   │   │       └── from-source.fmf
│   │   │   └── with-tmt-1.tgz
│   │   └── tests
│   │       ├── demo.spec
│   │       ├── .fmf
│   │       │   └── version
│   │       ├── from-source.txt
│   │       ├── mock_sources
│   │       ├── sources
│   │       ├── tests
│   │       │   └── from-source.fmf
│   │       ├── top_test.fmf
│   │       └── with-tmt-1.tgz
│   ├── step.yaml
│   └── tests.yaml
├── execute
│   ├── results.yaml
│   └── step.yaml
├── finish
│   └── step.yaml
├── prepare
│   ├── results.yaml
│   └── step.yaml
├── provision
│   ├── default-0
│   │   └── podman-run-environment
│   ├── guests.yaml
│   └── step.yaml
├── report
│   └── step.yaml
└── tree
    ├── demo.spec
    ├── .fmf
    │   └── version
    ├── mock_sources
    ├── sources
    ├── top_test.fmf
    └── with-tmt-1.tgz

24 directories, 38 files
```

</details>

and the archive content

<details>

```console
$ tar -tf with-tmt-1.tgz 
with-tmt-1/
with-tmt-1/from-source.txt
with-tmt-1/.fmf/
with-tmt-1/.fmf/version
with-tmt-1/tests/
with-tmt-1/tests/from-source.fmf
outsider
```

</details>

The issue here is we have 2 paths:
- `dist-git-merge` which should be the same result as the current after: it takes the inner tree and merges it in the test tree
- without `dist-git-merge` use the tree under the `source`

The old approach basically had an empty test tree so the same copy operation that `dist-git-merge` was "valid", but in the current PR it behaves like `dist-git-merge`.

A quick fix would be in the non-`dist-git-merge` to copy/symlink the whole source, and change `path` to be prefixed by the discovered dist-git fmf tree.

---

Long-term I would instead want to clean this up as:
- `test_dir` is always a symlink to the test tree
- redefine `source` to be the git cloned (dist-git source is the same here), unarchived path
- construct the symlink that points to the inner path under `source`

@therazix sounds good to you

File: tests/discover/distgit.sh [hunk 1]
```diff
@@ -439,9 +441,11 @@ EOF
...
         rlAssertExists $WORKDIR_SOURCE/outsider
 
-        # Test dir has only fmf_root from source (so one less level)
-        rlAssertExists $WORKDIR_TESTS/tests/from-source.fmf
-        rlAssertNotExists $WORKDIR_TESTS/with-tmt-1/tests/from-source.fmf
+        # The path structure is not flattened
+        # TODO: The behavior here for these first 2 checks has flipped this may have
+        #  some unexpected behavior, but seems very edge-case scenario
+        rlAssertNotExists $WORKDIR_TESTS/tests/from-source.fmf
+        rlAssertExists $WORKDIR_TESTS/with-tmt-1/tests/from-source.fmf
```
Comment (LecrisUT) [inline tests/discover/distgit.sh:448]: @lukaszachy what do you think about having this behavior change? I've also thought of a potentially more robust way to handle these (for good? 🤞) in the other comment https://github.com/teemtee/tmt/pull/4973#issuecomment-4660747605

Basically flattening or not would be controlled by `dist-git-merge`

Comment (lukaszachy) [reply tests/discover/distgit.sh:448]: @LecrisUT Thank you for working on this. I believe this is an acceptable breaking change. Such edge cases shouldn't be used much.

Also +1 to the proposal in that comment. The original dist-git implementation was overengineered for too little benefit.

**Comment by thrix** [comment]
@LecrisUT Can we add a release note just to be covered if somebody would be poking why there was a behaviour change pls?

File: tmt/steps/discover/fmf.py [hunk 2]
```diff
@@ -870,6 +856,12 @@ def post_dist_git(self, created_content: list[Path]) -> None:
                 raise tmt.utils.DiscoverError(
                     f"Couldn't glob '{dist_git_extract}' within extracted sources."
                 ) from error
+            if not dist_git_merge:
+                if self.data.path:
+                    self._logger.warning(
```
Comment (thrix) [inline tmt/steps/discover/fmf.py]: @LecrisUT rest of the module uses `self.warn`, why to use `self._logger.warning` instead here pls?

Comment (LecrisUT) [reply tmt/steps/discover/fmf.py]: API difference, did not port the `sources` and `reason` to these. Wanted to add them, but walked back on it, will do that another time

File: tmt/steps/discover/fmf.py [hunk 3]
```diff
@@ -604,29 +605,20 @@ def _fetch_local_repository(self) -> Optional[Path]:
 
         # Path for distgit sources cannot be checked until
         # they are extracted
```
Comment (thrix) [inline tmt/steps/discover/fmf.py]: This comment would now maybe deserve an update?

**Review by thrix — DISMISSED** [review]

File: tmt/steps/discover/fmf.py (hunk 2 above)
Comment (thrix) [reply tmt/steps/discover/fmf.py]: ok, I can live with it :)

File: tmt/steps/discover/fmf.py [hunk 4]
```diff
@@ -870,6 +856,12 @@ def post_dist_git(self, created_content: list[Path]) -> None:
                 raise tmt.utils.DiscoverError(
                     f"Couldn't glob '{dist_git_extract}' within extracted sources."
                 ) from error
+            if not dist_git_merge:
+                if self.data.path:
+                    self._logger.warning(
+                        "Automatic fmf root detection combined with 'path' key is ill defined"
```
Comment (bajertom) [inline tmt/steps/discover/fmf.py]: This tells the user "you did it wrong", but does not explain how to make it right or what would be the outcome of the ill defined behavior.

Comment (LecrisUT) [reply tmt/steps/discover/fmf.py]: The weird situation that we are in is that it is ill defined on our side. I do not think it is widely used though (only some of my packages). After https://github.com/teemtee/tmt/pull/4973#issuecomment-4660747605 we could tighten it up.

In the meantime, any better wording that could be used here?

Comment (thrix) [reply tmt/steps/discover/fmf.py]: Yeah, I would expand on it

Comment (LecrisUT) [reply tmt/steps/discover/fmf.py]: > Yeah, I would expand on it

How though? It is intentionally vague, because

> The weird situation that we are in is that it is ill defined on our side

I do not know what `path` is meant to signify and we have not known for a while (https://github.com/teemtee/tmt/issues/2097).

Comment (bajertom) [reply tmt/steps/discover/fmf.py]: > The weird situation that we are in is that it is ill defined on our side. 

Ah, in that case, seems OK, especially with the addition from last commit.

File: docs/releases/pending/4973.fmf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+description: |
+  Changed the location where the dist-git tests are extracted to when not using
+  ``dist-git-merge``. This may impact  the discovery when used with ``path``.
```
Comment (psss) [inline docs/releases/pending/4973.fmf]:
```suggestion
  ``dist-git-merge``. This may impact the discovery when used with ``path``.
```

File: docs/releases/pending/4973.fmf [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+  .. note::
```
Comment (psss) [inline docs/releases/pending/4973.fmf]: So far we haven't been using `note` directives in release notes. Does it really need such a [prominent place](https://tmt--4973.org.readthedocs.build/en/4973/releases/index.html#pending)? Maybe a bold warning would be just enough?

File: tmt/steps/discover/fmf.py [hunk 7]
```diff
@@ -602,31 +603,20 @@ def _fetch_local_repository(self) -> Optional[Path]:
...
+        if not git_root:
+            self.debug(f"Git root not found, using '{fmf_root}.'")
+            git_root = fmf_root
+        # Set path to relative path from the git root to fmf root
+        path = fmf_root.resolve().relative_to(
+            git_root.resolve() if requires_git else fmf_root.resolve()
+        )
 
         # Copy the git/fmf root directory to test_dir
         # (for dist-git case only when merge explicitly requested)
```
Comment (psss) [inline tmt/steps/discover/fmf.py]: Second line of this comment now seems to be outdated.

**Review by psss — COMMENTED** [review]
Looks good, just three minor nitpicks.

File: docs/releases/pending/4973.fmf (hunk 6 above)
Comment (psss) [reply docs/releases/pending/4973.fmf]: Does not seem to be [rendered corectly](https://tmt--4973.org.readthedocs.build/en/4973/releases/index.html#pending).

File: tmt/steps/discover/fmf.py (hunk 7 above)
Comment (psss) [reply tmt/steps/discover/fmf.py]: The first line is actually useful.

File: docs/releases/pending/4973.fmf [hunk 8]
```diff
@@ -0,0 +1,5 @@
+description: |
+  Changed the location where the dist-git tests are extracted to when not using
+  ``dist-git-merge``. This may impact the discovery when used with ``path``.
+
+  _Further changes and simplifications to these paths are planned._
```
Comment (psss) [inline docs/releases/pending/4973.fmf]:
```suggestion
  *Further changes and simplifications to these paths are planned.*
```

Underscores are not rendered in rst, did you mean `*` for italics?

**Review by psss — APPROVED** [review]

File: docs/releases/pending/4973.fmf (hunk 8 above)
Comment (LecrisUT) [reply docs/releases/pending/4973.fmf]: Yeah, wanted non-shouty italics. rst will be the bane of me

File: tmt/steps/discover/fmf.py [hunk 9]
```diff
@@ -913,10 +904,19 @@ def post_dist_git(self, created_content: list[Path]) -> None:
                     self.debug(f"Copy '{git_root}' to '{self.test_dir}'.")
                     if not self.is_dry_run:
                         tmt.utils.filesystem.copy_tree(git_root, self.test_dir, self._logger)
+            else:
+                if not dist_git_merge:
+                    if self.data.path:
+                        self.warn(
+                            "Automatic fmf root detection combined with 'path' key is ill"
+                            "defined. "
```
Comment (bajertom) [inline tmt/steps/discover/fmf.py]: LGTM, just this nitpick - this line would print "..key is illdefined. This.."

**Review by bajertom — APPROVED** [review]

**Comment by happz** [comment]
AFAICT, the distgit issue is gone, remaining failures are unrelated to this PR, and will be fixed in separate PRs (already in progress...)
